nselib/idna.lua

```diff
@@ -0,0 +1,614 @@
+local char = string.char
+local find = string.find
+local match = string.match
+local reverse = string.reverse
+local sub = string.sub
+
+-- This function concatenates the strings and tables (depth = 1) in
+-- a given table.
+--
+-- @param tbl A table is given as an input which contains values as string
+-- or table (depth = 1).
+-- @return Returns table after concatinating all the values.
+local function concat_table_in_tables(tbl)
+
+  local t = {}
+  for _, v in ipairs(tbl) do
+    if type(v) == "table" then
+      for _, q in ipairs(v) do
+        table.insert(t, q)
+      end
+    else
+      table.insert(t, v)
+    end
+  end
+
+  return t
+
+end
+
+-- This function breaks the tables of codepoints using a delimiter.
+--
+-- @param A table is given as an input which contains codepoints.
+-- @param ASCII value of delimiter is provided.
+-- @return Returns table of tables after breaking the give table using delimiter.
+function breakInput(codepoints, delimiter)
+
+  local tbl = {}
+  local output = {}
+
+  local delimiter = delimiter or 0x002E
+
+  for _, v in ipairs(codepoints) do
+    if v == delimiter then
+      table.insert(output, tbl)
+      tbl = {}
+    else
+      table.insert(tbl, v)
+    end
+  end
+
+  table.insert(output, tbl)
+
+  return output
+
+end
+
+-- This function maps the codepoints of the input to their respective
+-- codepoints based on the latest IDNA version mapping.
+--
+-- @param decoded_tbl Table of Unicode decoded codepoints.
+-- @param useSTD3ASCIIRules Boolean value to set the mapping according to IDNA2003 rules.
+--        useSTD3ASCIIRules=true refers to IDNA2008.
+--        useSTD3ASCIIRules=false refers to IDNA2003.
+-- @param transitionalProcessing Processing option to handle deviation codepoints.
+--        transitionalProcessing=true maps deviation codepoints to the input.
+--        transitionalProcessing=false maintains original input.
+-- @param viewDisallowedCodePoints Boolean value to see the list of disallowed codepoints.
+-- @return Returns table with the list of mapped codepoints.
+function map(decoded_tbl, useSTD3ASCIIRules, transitionalProcessing, viewDisallowedCodePoints)
+
+  -- Assigns default values if not specified.
+
+  -- According to IDNA2008, transitionalProcessing=true (default).
+  if transitionalProcessing == nil then
+    transitionalProcessing = true
+  end
+
+  if useSTD3ASCIIRules == nil then
+    useSTD3ASCIIRules = true
+  end
+  if viewDisallowedCodePoints == nil then
+    viewDisallowedCodePoints = false
+  end
+
+  local disallowedCodePoints = {}
+
+  if transitionalProcessing then
+    for index, cp in ipairs(decoded_tbl) do
+      local lookup = idnaMappings[cp]
+      if type(lookup) == "table" then
+        if lookup.status == "deviation" then
+          decoded_tbl[index] = lookup[1]
+        end
+      end
+    end
+  end
+
+  decoded_tbl = concat_table_in_tables(decoded_tbl)
+
+  -- Regular expressions (RFC 3490 separators)
+  for index, cp in ipairs(decoded_tbl) do
```

Comment:
These are already handled by `idnaMappings`, so no need to check separately for them.

Reply:
Removed this piece of code.